### PR TITLE
Add Continuous Delivery Jenkinsfile 

### LINKED ISSUE #280: Provide Continuous Delivery for TestGrid
**Description:**
We need to update the [TestGrid live deployment](http://ec2-34-238-28-168.compute-1.amazonaws.com:8080/) whenever a new binary become available at https://wso2.org/jenkins/job/platform-builds/job/wso2-test-grid/

We can directly update the production deployment until we have the pre-prod and dev environments ready. When we have the pre-prod, Jenkins continuous delivery should first update the dev. After a round of verification, we can move the changes to pre-prod and then to prod.

See the workflow below:

testgrid-deployment-continuous-delivery

I think we can make this work via a Jenkins pipeline.

1. Build TestGrid
If [build is success], then,
2. SCM tag it, and upload artifacts to nexus.
3. Do the deployment to TestGrid-Live Dev deployment
4. Wait for user-input on whether to do the deployment on TestGrid-Live Pre-Prod.
5. Similarly, wait for user-input on whether to do the deployment on TestGrid-Live PROD.

### PR DESCRIPTION
**Pull request by sameerawickramasekara**
## Goals
This script will create a Jenkins job that can provide continuous delivery for TestGrid.

## Approach
untitled diagram 2

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: jenkins-pipelines/continous-delivery-jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                sshagent(['//Add Dev enviromnet key ID']) {
+                    sh """
+          		        scp web/target/*.war ${PROD_USER}@${PROD_HOST}:~/
+          	        """
+                }
+            }
+
+        }
+    }
+}
```
Comment (yasassri) [inline jenkins-pipelines/continous-delivery-jenkinsfile]: Should we add a newline?

File: jenkins-pipelines/continueous-delivery-jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,62 @@
+pipeline {
+    agent any
+    environment {
+
+        DEV_USER = 'xxxxx'
+        DEV_HOST = 'xxxxxxx.com'
+        PRE_PROD_USER = 'xxxxx'
+        PRE_PROD_HOST = 'xxxxxx.com'
+        PROD_USER = 'xxxxx'
+        PROD_HOST = 'xxxxxxx.com'
```
Comment (kasunbg) [inline jenkins-pipelines/continueous-delivery-jenkinsfile]: We should inject these values via envInject plugin. May be setting envInject can over-ride the environment variables set in this file.

File: jenkins-pipelines/continueous-delivery-jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    stages {
+        stage('Deploy to Dev') {
+            steps {
+                copyArtifacts(projectName: 'testgrid-build');
```
Comment (kasunbg) [inline jenkins-pipelines/continueous-delivery-jenkinsfile]: Please add a comment mentioning which files u actually get as part of this process.

File: jenkins-pipelines/continueous-delivery-jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                sshagent(['//Add Dev enviromnet key ID']) {
```
Comment (kasunbg) [inline jenkins-pipelines/continueous-delivery-jenkinsfile]: Assume this is testgrid-dev-key

File: jenkins-pipelines/continueous-delivery-jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    sh """
+                		   scp web/target/*.war ${DEV_USER}@${DEV_HOST}:~/
+
+                	    """
+                }
+
+                sshagent(['//Add Dev enviromnet key ID'']) {
+                    sh """
+                 		   scp distribution/target/*.zip ${DEV_USER}@${DEV_HOST}:~/
```
Comment (kasunbg) [inline jenkins-pipelines/continueous-delivery-jenkinsfile]: We can merge this sshAgent section with the one above.

File: jenkins-pipelines/continueous-delivery-jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            if (userInput == true) {
+                sshagent(['//ADD Prod Environment Key ID']) {
+                    sh """
+                     scp distribution/target/*.zip ${PROD_USER}@${PROD_HOST}:~/
+
+         	        """
+                }
+                sshagent(['//ADD Prod Environment Key ID']) {
+                    sh """
+          		        scp web/target/*.war ${PROD_USER}@${PROD_HOST}:~/
```
Comment (kasunbg) [inline jenkins-pipelines/continueous-delivery-jenkinsfile]: merge with above.

**Pull request merged**